File: assignment/02/journey_to_the_moon.cc
// alternative (BFS) provided by TA 鄧晉杰
long journeyToMoon (long n, std::vector<std::vector<long>> astronaut)
{
  std::vector<std::vector<long>> graph(n);
  for (auto pair : astronaut)
  {
    graph[pair[0]].emplace_back(pair[1]);
    graph[pair[1]].emplace_back(pair[0]);
  }
  std::vector<bool> is_discovered(n);
  long amount_valid_pairs {};
  long country_size_sum {};
  for (long id {}; id != n; ++id)
  {
    if (!is_discovered[id])
    {
      long country_size {};
      std::deque<long> astronauts;
      is_discovered[id] = true;
      astronauts.emplace_back(id);
      while (!astronauts.empty())
      {
        ++country_size;
        auto u {astronauts.front()};
        astronauts.pop_front();
        for (auto v : graph[u])
        {
          if (!is_discovered[v])
          {
            is_discovered[v] = true;
            astronauts.emplace_back(v);
          }
        }
      }
      amount_valid_pairs += (country_size_sum * country_size);
      country_size_sum += country_size;
    }
  }
  return amount_valid_pairs;
}

File: assignment/02/journey_to_the_moon.py
# alternative provided by TA 蔡宗霖
def BFS(graph, find, start, weight):
    cost = 0
    member = 1
    queue = deque([])
    find[start] = True
    queue.append(start)
    while len(queue) > 0:
        v = queue.popleft()
        cost += weight[v]
        for nei in graph[v]:
            if not find[nei]:
                find[nei] = True
                queue.append(nei)
                member += 1
    return(cost, member)


def journeyToMoon(n, astronaut):
    adj_l = [[] for _ in range(n)]
    weight = [0] * n
    find = [False] * n

    for s, t in astronaut:
        adj_l[s].append(t)
        adj_l[t].append(s)

    group = []

    for i in range(n):
        if not find[i]:
            cost, members = BFS(adj_l, find, i, weight)
            group.append(members)

    output = 0
    count = sum(group)
    for i in group[:-1]:
        count -= i
        output = output + (i * count)
    return(output)
